apiserver/facades/agent/upgradeseries/upgradeseries.go

```diff
@@ -121,6 +121,32 @@ func (a *API) SetMachineStatus(args params.UpgradeSeriesStatusParams) (params.Er
 	return result, nil
 }
 
+// CurrentSeries returns what Juju thinks the current series of the machine is.
+// Note that a machine could have been upgraded out-of-band by running
+// do-release-upgrade outside of the upgrade-series workflow,
+// making this value incorrect.
+func (a *API) CurrentSeries(args params.Entities) (params.StringResults, error) {
```

Comment:
The implementations of `CurrentSeries` and `TargetSeries` are identical with the exception of the method invoked on the machine to get the series. You could extract the common logic into a function such as:

```
type seriesGetterFn func(state.Machine) (string, error)
func resolveSeries(args params.Entities, seriesGetter seriesGetterFn) {
   ...
   res, err := seriesGetter(machine)
   ... 
}
```

Reply:
They aren't though. `Series` returns string, `UpgradeSeriesTarget` returns (string, error).

Reply:
Wouldn't this work since the public methods pass their own getter func?

```go
resolveSeries(args, func(m *state.Machine) (string, error) {
  return m.Series(), nil
})

resolveSeries(args, func(m *state.Machine) (string, error) {
  return m.UpgradeSeriesTarget()
})
```

Reply:
If there were another method in this vein, I would do this. 

I think the methods are sufficiently different (despite semantic similarity) - one querying the machine and the other the upgrade-series lock - that it's clearer to suffer a little duplication for clarity.

Reply:
Fair enough